sql/row: minor refactorings of the FK existence check code to add clarity

knz opened the pull request:
I'm currently prototyping something and I had to reverse-engineer the code. As a result here are some code movement into separate files and many explanatory comments.

Here is a list of mis-designs and inefficiencies I have discovered while doing so:

- `fkExistenceCheckBaseHelper` and `fkExistenceBatchChecker` store the `*client.Txn`; this should be passed around as argument instead (like `context.Context`).
- `fkExistenceCheckBaseHelper` stores the entire `sqlbase.ForeignKeyReference` object, whereas it only needs the match style. This can be simplified.
- `(f *fkExistenceBatchChecker) runCheck()` re-allocated Datum slices for every row. The slice should be pre-allocated and reused.
- `queueFkExistenceChecksForRow()` performs the same check over and over on the match style, whereas the match style is fixed for an entire constraint. This should be hoisted to avoid a per-row overhead.
- `MakeFkMetadata()` instantiates a `*CheckHelper` for every table, even when it is not needed and there are no CASCADE actions. This is inefficient; the check helpers should only be instantiated when needed.
- the `fkExistenceCheckForUpdate` object is unable to determine itself the set of FK constraints it should check and instead relies on its caller to provide this information. This seems incorrect/error-prone.
- the `fkExistenceCheckForUpdate` object is abused to also determine *whether* there are FK constraints to be checked during an update. This is improper abuse of something designed for an unrelated purpose. The boolean `hasFk()` should be hoisted and determined in the caller instead.
- `(fks fkExistenceCheckForUpdate) addCheckForIndex()` skips over inverted indexes, but there is no explanation as to why. This needs to be verified and documented.
- the existence of the `FkTablesMetadata` is redundant with schema access services already available in packages `sql` and `sqlbase`. There must be a way to factor them and avoid this redundancy.
- throughout: the FK constraint definitions are much too tied to index definitions.

There is a possible problem with the entire FK code simply ignoring the FK work (either existence checks or CASCADE actions) if the table(s) involved in constraints is not public or in the process of being added. This feels wrong and needs to be investigated + clarified.

knz committed:
sql/row: split the FK span generation into a separate file

Release note: None

knz committed:
sql/row: move the base FK to a separate file, add docs

Release note: None

knz committed:
sql/row: use pgerror.NewAssertionErrorf

Release note: None

knz committed:
sql/row: clarify the initialization of the base FK helper

Release note: None

knz committed:
sql/*: properly use IndexDescriptor references

The API definition in `sqlbase` for `FindIndexByID` and
`AllNonDropIndexes` was tortured and mangled into passing
IndexDescriptors by value. This is insane as it 1) requires heap
allocations up to 2 times for each API use 2) causes a lot of memory
traffic, as these structs are not trivial in size.

Since the index descriptors are always on the heap to start with, we
may as well use them by reference all the time.

Also this simplifies the code!!

Release note: None

knz committed:
sql/row: rename writeIdx -> mutatedIdx

Release note: None

knz committed:
sql/row: move the low-level FK existence checker to its own file

Release note: None

knz committed:
sql/row: move the fkInsertHelper to its own file

Release note: None

knz committed:
sql/row: move fkDeleteHelper/fkUpdateHelper to separate files

Additionally, there was some unneeded cruft in the structs, which I
removed.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/34094)
<!-- Reviewable:end -->

knz committed:
sql/row: add clarifying comments.

Release note: None

BramGruneir reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

LGreatTM 

Thank you so much for making these changes!   

Just one spelling error in the final commit.

> Reviewed 2 of 2 files at r1, 2 of 2 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 13 of 13 files at r5, 2 of 2 files at r6, 2 of 2 files at r7, 2 of 2 files at r8, 3 of 3 files at r9, 1 of 1 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34094#-:-LWSXDza9YvFcLRytv5B:bbqyyr9)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/row/fk_base_helper.go, line 12 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34094#-LWSTttQDZLGf4VRr9Vi:-LWSTttQDZLGf4VRr9Vj:b-1n10yq) ([raw file](https://github.com/cockroachdb/cockroach/blob/eefe876e068eb64a75de0882475e1ff3fbc2bed7/pkg/sql/row/fk_base_helper.go#L12)):*
> ```Go
> )
> 
> // baseFKHelper is an auxiliary struct that facilitates FK existence
> ```

I LOVE these comments!  
___
*[pkg/sql/row/fk_base_helper.go, line 120 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/34094#-LWSV1LpC-_ss1rCz35a:-LWSV1LpC-_ss1rCz35b:b-lfij1g) ([raw file](https://github.com/cockroachdb/cockroach/blob/35432134c6e60877969e701e56dce385539b328d/pkg/sql/row/fk_base_helper.go#L120)):*
> ```Go
> 
> 	// Determine the number of columns being looked up.
> 	prefixLen := len(searchIdx.ColumnIDs)
> ```

This logic seems odd.  Can we not pull it directly from ref?
___
*[pkg/sql/row/fk_batch_checker.go, line 107 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/34094#-LWSWzH6502nv3dhsjCf:-LWSWzH6502nv3dhsjCg:b2qpvf4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d30fea85f90026275b3ada5bee2eda2faef1b79/pkg/sql/row/fk_batch_checker.go#L107)):*
> ```Go
> 			if fk.rf.kvEnd {
> 				// TODO(knz): re-allocating a datum slice in every check
> 				// is super inefficient and expensive. Factor this.
> ```

Factor should be Refactor.
___
*[pkg/sql/row/fk_batch_checker.go, line 128 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/34094#-LWSX1V43aCEMCr_i6C3:-LWSX1V43aCEMCr_i6C4:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d30fea85f90026275b3ada5bee2eda2faef1b79/pkg/sql/row/fk_batch_checker.go#L128)):*
> ```Go
> 
> 				// TODO(knz): re-allocating a datum slice in every check
> 				// is super inefficient and expensive. Factor this.
> ```

ditto


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34094)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @bobvawter, @BramGruneir, and @knz)

___
*[pkg/sql/row/fk_base_helper.go, line 120 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/34094#-LWSV1LpC-_ss1rCz35a:-LWWI_EU2lPzX9pDQGHp:bh3zzws) ([raw file](https://github.com/cockroachdb/cockroach/blob/35432134c6e60877969e701e56dce385539b328d/pkg/sql/row/fk_base_helper.go#L120)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

This logic seems odd.  Can we not pull it directly from ref?
</blockquote></details>

I don't believe so, no. See the comment on `prefixLen` in the baseFKHelper.
___
*[pkg/sql/row/fk_batch_checker.go, line 107 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/34094#-LWSWzH6502nv3dhsjCf:-LWWIeeP2MN5u41ZhsrF:b-jtvhrs) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d30fea85f90026275b3ada5bee2eda2faef1b79/pkg/sql/row/fk_batch_checker.go#L107)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Factor should be Refactor.
</blockquote></details>

actually, in this specific case, no. The word "re-factoring" means doing the "factoring" a new time. Factoring is the process of making things that are common to multiple places written just once (DRY). 

Here making the alloc occuring one, is a factoring change. If we then change the location of this combined alloc to a different place, that will be a re-factoring. But we need to factor first.

See also https://lispcast.com/stop-refactoring-and-start-factoring/


<!-- Sent from Reviewable.io -->

knz committed:
sql/row: clarify the code by renaming some objects

- "fk helper" -> "fk existence check helper"
- `checkIdx` -> `queueFkExistenceChecksForRow`

Release note: None

knz committed:
sql/row: move an error def

Release note: None

knz committed:
sql/row: rename fk.go -> fk_existence_prepare.go

Release note: None

knz committed:
sql/row: move the table lookup logic to a separate file; add comments

Release note: None

knz committed:
sql/row: minor renames for clarity

Release note: None

knz committed:
sql/row: add a missing comment

Release note: None

knz committed:
sql/row: rename row.TableLookup struct -> row.TableEntry

Release note: None

knz committed:
sql/row: minor renames to clarify; add comments

Release note: None

knz committed:
sql/row: minor renames to improve readability

- "TableLookupsByID" -> "FkTableMetadata"
- "ID" -> "TableID"

Release note: None

knz committed:
sql: clarify some code

Release note: None

knz committed:
sql/row: minor rename to improve readability

Release note: None

knz committed:
sql/row: rename TablesNeededForFKs -> MakeFkMetadata

Release note: None

knz committed:
sql/row: add a clarifying comment

Release note: None

knz commented:
bors r+
